Enable serialization and parsing from xContent for SmoothingModels

cbuescher opened the pull request:
PhraseSuggestionBuilder uses three smoothing models internally. In order to enable proper serialization / parsing from xContent to the phrase suggester for the search builder refactoring, this change starts by making the smoothing models writable, adding hashCode/equals and fromXContent.

areek commented on core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java:
```diff
@@ -313,40 +377,111 @@ protected XContentBuilder innerToXContent(XContentBuilder builder, Params params
+    }
+
+
+    public static abstract class SmoothingModel<SM extends SmoothingModel<?>> implements NamedWriteable<SM>, ToXContent {
```
I don't understand why we need to add the type parameter to `SmoothingModel`, I have taken a stab at removing it [here](https://gist.github.com/areek/e83003e235011c8ebd48). IMO, a non-generic `SmoothingModel` is simpler. WDYT?

cbuescher replied:
I'm absolutely okay to remove it. I think I just added it because we used this for the abstract query builder and some other elements in past refactorings to avoid some casting, but I'm also glad to simlify it here. Will change this.

cbuescher commented:
@areek thanks for the review, I removed the type parameter from SmoothingModel, also added an aditional buildWordScorerFactory() method that we will need later in the Phrase suggesters build method. Can you take another look if this now look okay to you?

areek commented on core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java:
```diff
@@ -376,19 +536,107 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            XContentParser parser = parseContext.parser();
+            XContentParser.Token token;
+            String fieldName = null;
+            final double[] lambdas = new double[3];
```
Nit: would be nice if these were explicit variables, e.g. `trigramLambda`, `bigramLambda` and `unigramLambda` instead?

cbuescher replied:
Sure, pure lazyness copied over from current PhraseSuggestParser. Will change that.

areek commented:
@cbuescher this looks great, LGTM! I added a minor style comment
